### PR TITLE
Check image locally before pull

### PR DESCRIPTION
**Pull request by superseb**
https://github.com/rancher/rke/issues/137

### CODE REVIEW

File: docker/docker.go [hunk 1]
```diff
@@ -130,6 +130,21 @@ func IsContainerRunning(dClient *client.Client, hostname string, containerName s
 	return false, nil
 }
 
+func LocalImageExists(dClient *client.Client, hostname string, containerImage string) (bool, error) {
```
Comment (galal-hussein) [inline docker/docker.go]: i don't think this function should be exported

File: docker/docker.go [hunk 2]
```diff
@@ -64,12 +64,12 @@ func DoRollingUpdateContainer(dClient *client.Client, imageCfg *container.Config
...
 	}
-	logrus.Infof("[%s] Pulling Image on host [%s]", plane, hostname)
-	err = PullImage(dClient, hostname, imageCfg.Image)
+	logrus.Infof("[%s] Checking image on host [%s]", plane, hostname)
+	err = UseLocalOrPull(dClient, hostname, imageCfg.Image)
 	if err != nil {
 		return err
 	}
-	logrus.Infof("[%s] Successfully pulled [%s] image on host [%s]", plane, containerName, hostname)
+	logrus.Infof("[%s] Image [%s] exists on host [%s]", plane, containerName, hostname)
```
Comment (galal-hussein) [inline docker/docker.go]: this info message will show that image exist even if it pulled the image, it should reflect what happened exactly (pulled or found)

Comment (galal-hussein) [reply docker/docker.go]: maybe add the message in UseLocalOrPull function, instead of using it for each occurrence

File: docker/docker.go [hunk 3]
```diff
@@ -130,6 +130,21 @@ func IsContainerRunning(dClient *client.Client, hostname string, containerName s
...
 
+func LocalImageExists(dClient *client.Client, hostname string, containerImage string) (bool, error) {
+	logrus.Debugf("Checking if image [%s] exists on host [%s]", containerImage, hostname)
+	_, _, err := dClient.ImageInspectWithRaw(context.Background(), containerImage)
+	if err != nil {
+		if client.IsErrNotFound(err) {
+			logrus.Debugf("Image [%s] does not exist on host [%s]: %v", containerImage, hostname, err)
+			return false, nil
+		}
+		logrus.Debugf("Error checking if image [%s] exists on host [%s]: %v", containerImage, hostname, err)
```
Comment (galal-hussein) [inline docker/docker.go]: there is no need for extra debug message, its the same as the error returned

File: pki/deploy.go [hunk 4]
```diff
@@ -62,8 +62,8 @@ func DeployCertificatesOnWorkers(workerHosts []*hosts.Host, crtMap map[string]Ce
 }
 
 func doRunDeployer(host *hosts.Host, containerEnv []string, certDownloaderImage string) error {
-	logrus.Debugf("[certificates] Pulling Certificate downloader Image on host [%s]", host.Address)
-	if err := docker.PullImage(host.DClient, host.Address, certDownloaderImage); err != nil {
+	logrus.Debugf("[certificates] Checking image [%s] on host [%s]", certDownloaderImage, host.Address)
```
Comment (galal-hussein) [inline pki/deploy.go]: same as before, can you remove the `Checking image` logs from all calls and add it instead to the UseLocalOrPull

File: docker/docker.go (hunk 2 above)
Comment (superseb) [reply docker/docker.go]: The reasoning here is that you would lose `plane` in the log message when you move it to `UseLocalOrPull`. Passing `plane` for just the log doesn't seem the right way, even as printing the message as return from the function.

File: docker/docker.go [hunk 5]
```diff
@@ -145,6 +155,24 @@ func PullImage(dClient *client.Client, hostname string, containerImage string) e
 	return nil
 }
 
+func UseLocalOrPull(dClient *client.Client, hostname string, containerImage string, plane string) error {
+	logrus.Infof("[%s] Checking image [%s] on host [%s]", plane, containerImage, hostname)
+	imageExists, err := localImageExists(dClient, hostname, containerImage)
+	if err != nil {
+		return err
+	}
+	if !imageExists {
```
Comment (alena1108) [inline docker/docker.go]: Nit: better to use if imageExists { return nil} construct. Always better to check the positive

File: docker/docker.go [hunk 6]
```diff
@@ -145,6 +155,24 @@ func PullImage(dClient *client.Client, hostname string, containerImage string) e
...
 }
 
+func UseLocalOrPull(dClient *client.Client, hostname string, containerImage string, plane string) error {
+	logrus.Infof("[%s] Checking image [%s] on host [%s]", plane, containerImage, hostname)
+	imageExists, err := localImageExists(dClient, hostname, containerImage)
+	if err != nil {
+		return err
+	}
+	if imageExists {
+		logrus.Infof("[%s] No pull necessary, image [%s] exists on host [%s]", plane, containerImage, hostname)
```
Comment (alena1108) [inline docker/docker.go]: we should log it in debug

**Pull request merged**